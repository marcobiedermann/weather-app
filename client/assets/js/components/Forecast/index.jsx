import PropTypes from 'prop-types';
import React from 'react';
import Weather from '../Weather';

const Forecast = props => {
  const { list, ...otherProps } = props;

  return (
    <div {...otherProps}>
      {list && (
        <ul>
          {list.map(weather => (
            <li key={weather.dt}>
              <Weather {...weather} />
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

Forecast.propTypes = {
  list: PropTypes.arrayOf(PropTypes.shape()),
};

Forecast.defaultProps = {
  list: [],
};

export default Forecast;
