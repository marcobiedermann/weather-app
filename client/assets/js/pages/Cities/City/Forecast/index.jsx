import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Helmet } from 'react-helmet';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { fetchForecastById } from '../../../../actions/forecast';

class ForecastPage extends Component {
  componentDidMount() {
    const { fetchForecastById, match } = this.props;

    fetchForecastById(match.params.cityId);
  }

  render() {
    const { forecast } = this.props;

    return (
      <div>
        <Helmet>
          <title>Forecast</title>
        </Helmet>
        <ul>
          {forecast.list.map((item) => (
            <li key={item.dt}>
              <div>Min: {item.temp.min}</div>
              <div>Max: {item.temp.max}</div>
              <div>
                {item.weather[0].description}
                <img
                  src={`https://openweathermap.org/img/w/${item.weather[0].icon}.png`}
                  alt={item.weather[0].description}
                />
              </div>
            </li>
          ))}
        </ul>
      </div>
    );
  }
}

ForecastPage.propTypes = {
  forecast: PropTypes.shape({
    list: PropTypes.arrayOf(
      PropTypes.shape({
        dt: PropTypes.number,
        temp: PropTypes.shape({
          min: PropTypes.number,
          max: PropTypes.number,
        }),
        weather: PropTypes.arrayOf(
          PropTypes.shape({
            description: PropTypes.string,
            icon: PropTypes.string,
          }),
        ),
      }),
    ),
  }),
  fetchForecastById: PropTypes.func,
  match: PropTypes.shape({
    params: PropTypes.shape({
      cityId: PropTypes.string,
    }),
    path: PropTypes.string,
  }),
};

ForecastPage.defaultProps = {
  forecast: null,
  fetchForecastById: () => {},
  match: null,
};

const mapStateToProps = (state) => ({
  forecast: state.forecast,
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      fetchForecastById,
    },
    dispatch,
  );

export default connect(mapStateToProps, mapDispatchToProps)(ForecastPage);
