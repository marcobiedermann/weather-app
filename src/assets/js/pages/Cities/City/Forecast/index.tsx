import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import useSWR from 'swr';
import Loader from '../../../../components/Loader';

interface Params {
  cityId: string;
}

interface ForecastPageQuery {
  city: City;
  cnt: number;
  cod: string;
  list: List[];
  message: number;
}

interface City {
  coord: Coord;
  country: string;
  id: number;
  name: string;
  population: number;
  timezone: number;
}

interface Coord {
  lat: number;
  lon: number;
}

interface List {
  clouds: number;
  deg: number;
  dt: number;
  feels_like: FeelsLike;
  humidity: number;
  pop: number;
  pressure: number;
  speed: number;
  sunrise: number;
  sunset: number;
  temp: Temp;
  weather: Weather[];
}

interface FeelsLike {
  day: number;
  eve: number;
  morn: number;
  night: number;
}

interface Temp {
  day: number;
  eve: number;
  max: number;
  min: number;
  morn: number;
  night: number;
}

interface Weather {
  description: string;
  icon: string;
  id: number;
  main: string;
}

const ForecastPage: FC = () => {
  const { cityId } = useParams<Params>();
  const { data, error } = useSWR<ForecastPageQuery>(`/forecast/daily?id=${cityId}`);

  if (error) {
    return <div>Error: {error}</div>;
  }

  if (!data) {
    return <Loader />;
  }

  return (
    <>
      <Helmet>
        <title>Forecast</title>
      </Helmet>
      <ul>
        {data.list.map((item) => (
          <li key={item.dt}>
            <div>Min: {item.temp.min}</div>
            <div>Max: {item.temp.max}</div>
            <div>
              {item.weather[0].description}
              <img
                src={`https://openweathermap.org/img/w/${item.weather[0].icon}.png`}
                alt={item.weather[0].description}
              />
            </div>
          </li>
        ))}
      </ul>
    </>
  );
};

export default ForecastPage;
