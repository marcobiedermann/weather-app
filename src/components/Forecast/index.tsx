import format from 'date-fns/format';
import fromUnixTime from 'date-fns/fromUnixTime';

interface Temp {
  max: number;
  min: number;
}

interface Weather {
  description: string;
  icon: string;
  id: number;
}
interface List {
  dt: number;
  temp: Temp;
  weather: Weather[];
}
interface ForecastProps {
  list: List[];
}

function Forecast(props: ForecastProps): JSX.Element {
  const { list } = props;

  return (
    <table>
      <tbody>
        {list.map((day) => (
          <tr key={day.dt}>
            <td>
              <time dateTime={format(fromUnixTime(day.dt), 'yyyy-MM-dd')}>
                {format(fromUnixTime(day.dt), 'EEEE')}
              </time>
            </td>
            <td>
              {day.weather.map((entry) => (
                <img
                  src={`https://openweathermap.org/img/wn/${entry.icon}.png`}
                  alt={entry.description}
                  key={entry.id}
                />
              ))}
            </td>
            <td>{Math.round(day.temp.max)} °C</td>
            <td>{Math.round(day.temp.min)} °C</td>
          </tr>
        ))}
      </tbody>
    </table>
  );
}

export type { ForecastProps };
export default Forecast;
