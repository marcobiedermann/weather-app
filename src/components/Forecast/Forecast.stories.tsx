import type { Meta, StoryObj } from '@storybook/react';
import Forecast from '.';

const meta: Meta<typeof Forecast> = {
  args: {
    list: [
      {
        dt: 1687100400,
        main: {
          temp: 25.34,
        },
        weather: [
          {
            description: 'scattered clouds',
            icon: '03d',
            id: 802,
          },
        ],
      },
    ],
  },
  component: Forecast,
};

type Story = StoryObj<typeof Forecast>;

const Default: Story = {};

export { Default };
export default meta;
