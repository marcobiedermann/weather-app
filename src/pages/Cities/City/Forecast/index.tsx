import React from 'react';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import Error from '../../../../components/Error';
import Forecast from '../../../../components/Forecast';
import Loader from '../../../../components/Loader';
import { useDailyForeCast } from '../../../../hooks';

function ForecastPage(): JSX.Element {
  const { cityId } = useParams();
  const { data, error, isError, isLoading } = useDailyForeCast(parseInt(cityId!, 10));

  if (isError) {
    return <Error message={error.message} />;
  }

  if (isLoading) {
    return <Loader />;
  }

  return (
    <>
      <Helmet>
        <title>Daily Forecast</title>
      </Helmet>
      <Forecast list={data?.list} />
    </>
  );
}

export default ForecastPage;
