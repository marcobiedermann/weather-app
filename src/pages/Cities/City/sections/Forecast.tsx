import { fromUnixTime, isSameDay } from 'date-fns';
import Error from '../../../../components/Error';
import Forecast from '../../../../components/Forecast';
import Loader from '../../../../components/Loader';
import { useForecast } from '../../../../hooks';

interface ForecastSectionProps {
  cityId: number;
  settings: any;
}

function ForecastSection(props: ForecastSectionProps) {
  const { cityId, settings } = props;
  const { data: forecast, error, isError, isLoading } = useForecast(cityId, settings);

  if (isError) {
    return <Error message={error.message} />;
  }

  if (isLoading) {
    return <Loader />;
  }

  const now = new Date();

  return (
    <section>
      <Forecast list={forecast.list.filter((day) => isSameDay(now, fromUnixTime(day.dt)))} />
    </section>
  );
}

export type { ForecastSectionProps };
export default ForecastSection;
