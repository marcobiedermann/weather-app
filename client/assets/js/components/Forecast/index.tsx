import React, { FC } from 'react';
import City, { CityProps } from '../City';

export interface ForecaseProps {
  list: CityProps[];
}

const Forecast: FC<ForecaseProps> = (props) => {
  const { list, ...otherProps } = props;

  return (
    <div {...otherProps}>
      {list && (
        <ul>
          {list.map((city) => (
            <li key={city.dt}>
              <City {...city} />
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

export default Forecast;
