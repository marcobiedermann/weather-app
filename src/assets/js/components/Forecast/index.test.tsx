import { render } from '@testing-library/react';
import React from 'react';
import Forecast from '.';

describe('components/Forecast', () => {
  it('renders correctly', () => {
    expect.assertions(1);

    const { asFragment } = render(<Forecast list={[]} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
