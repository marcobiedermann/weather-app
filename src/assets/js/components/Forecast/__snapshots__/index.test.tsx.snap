// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Forecast renders correctly 1`] = `
<DocumentFragment>
  <table>
    <tbody />
  </table>
</DocumentFragment>
`;
